---
title: "Library taxonomic analysis"
output: html_notebook
---

# run kraken.sh --> bracken.sh --> merge_bracken.sh

# plotting function for merged bracken data
```{r}
plot_bracken <- function(file, top = NA, pal, otherColor, sample_order){
    
    # file is the bracken data
    # top is an integer, trims the data to the number of top taxa (by median) in a given level for better plotting
    ## top = NA is the default and will plot all taxa in a given level
    # pal is a palette of hex colors
    # otherColor is the color used to represent taxa below median abundance threshold
    # sample_order is the ordered list of sample names used to determine x-axis ordering
    
    levdat <- file %>%
        select(-contains(c("_num","taxonomy_"))) %>%
        pivot_longer(!name, names_to = "sample", values_to = "fraction") %>%
        mutate(sample = str_remove_all(sample, "_frac"))
    
    level <- file %>%
        select(contains("taxonomy_lvl")) %>% unique()
    
    newpal <- str_split(pal, ",") %>% unlist() %>% paste0("#",.)
    
    if (level == "K") {
        label <- "Kingdom"
    } else if (level == "P") {
        label <- "Phylum"
    } else if (level == "C") {
        label <- "Class"
    } else if (level == "O") {
        label <- "Order"
    } else if (level == "F") {
        label <- "Family"
    } else if (level == "G") {
        label <- "Genus"
    } else if (level == "S") {
        label <- "Species"
    } else if (level == "S1") {
        label <- "Strain"
    } else {
        message("Bracken taxonomy_lvl needs to be K, P, C, O, F, G, S, or S1")
    }
    
    if (!is.na(top) && top < length(unique(levdat$name))) {
        
        top_select <- aggregate(levdat$fraction, by = list(levdat$name), FUN = median) %>%
            purrr::set_names(c("group","value")) %>%
            arrange(desc(value)) %>%
            dplyr::slice(1:top) %>% pull(var = group)
        
        levdat <- filter(levdat, levdat$name %in% top_select)
        
        fill_unk <- aggregate(levdat$fraction, by = list(levdat$sample), FUN = sum) %>% 
            purrr::set_names(c("sample","sum")) %>% 
            mutate_if(is.numeric, round, 3)
        
        fill_unk$residual <- 1 - fill_unk$sum
        
        for (i in 1:nrow(fill_unk)) {
            
            levdat <- rbind(levdat, c("other", fill_unk[i,1], fill_unk[i,3]))
            
        }
        
        levdat <- transform(levdat, fraction = as.numeric(fraction))
        levdat$name <- factor(levdat$name, levels = unique(levdat$name))
        levdat$sample <- factor(levdat$sample, levels = sample_order)
        
        fill_vals <- c(rep_len(newpal, nrow(unique(levdat[1]))-1), otherColor)
        
    } else {
        
        fill_vals <- c(rep_len(newpal, nrow(unique(levdat[1]))))
        
    }
    
    print(nrow(levdat))
    
    ggplot(data = levdat) +
        geom_bar(mapping = aes(x = sample, y = fraction, fill = name),
                 position = "fill", stat = "identity", width = 0.75) + 
        scale_fill_manual(paste(label),
                          values = fill_vals) +
        theme_classic() +
        scale_x_discrete("") +
        scale_y_continuous("Relative Abundance") +
        theme(axis.text.x = element_text(angle = -90, vjust = 0.5, hjust = 0, size = 11, color = "black"),
              axis.text.y = element_text(size = 11, color = "black"),
              axis.title.x = element_text(),
              axis.title.y = element_text(size = 13),
              legend.position = "right",
              axis.line = element_line(color = "black"))
    
    
    
}
```

# create plot for raw read data
```{r}
setwd("/workdir/users/acv46/stool_PROSeq3/kraken")

bdat <- read_tsv("bracken_merged_P.txt", col_names = TRUE) |>
  mutate(name = ifelse(name == "Firmicutes", "Bacillota", name),
         name = ifelse(name == "Bacteroidetes", "Bacteroidota", name),
         name = ifelse(name == "Actinobacteria", "Actinomycetota", name))
samples <- c(
             "US2_mgm",
             "US2_A_PROseq","US2_B_PROseq","US2_C_PROseq",
             "US2_A_RNAseq","US2_B_RNAseq",
             "US2_A_TEXm","US2_B_TEXm","US2_C_TEXm",
             "US2_A_TEXp","US2_B_TEXp","US2_C_TEXp",
             "US3_mgm",
             "US3_A_PROseq","US3_B_PROseq","US3_C_PROseq",
             "US3_A_RNAseq","US3_B_RNAseq",
             "US3_A_TEXm","US3_B_TEXm","US3_C_TEXm",
             "US3_A_TEXp","US3_B_TEXp","US3_C_TEXp"
             )

p1 <- plot_bracken(file = bdat,
                   top = 4,
                   pal = "a30000,f5b841,00798c,76e5fc,9acd32,8b008b,66cdaa",
                   otherColor = "gray",
                   sample_order = samples)

p1 <- p1 + geom_vline(xintercept = 12.5, color = "black")
p1
```

## ggsave
```{r}
setwd("/workdir/users/acv46/stool_PROSeq3/figures/kraken/")
ggsave(plot = p1,
       filename = "kraken_phylum_relative_abudance_v2.pdf",
       width = 7,
       height = 4,
       units = "in",
       device = cairo_pdf)
```

# create plot for aligned reads -- new data
```{r}
setwd("/workdir/users/acv46/stool_PROSeq3/kraken/merged_bracken_bams")

bdat <- read_tsv("bracken_merged_P.txt", col_names = TRUE) |>
  mutate(name = ifelse(name == "Firmicutes", "Bacillota", name),
         name = ifelse(name == "Bacteroidetes", "Bacteroidota", name),
         name = ifelse(name == "Actinobacteria", "Actinomycetota", name))
samples <- c(
             "US2_mgm",
             "US2_A_PROseq","US2_B_PROseq","US2_C_PROseq",
             "US2_A_RNAseq","US2_B_RNAseq",
             "US2_A_TEXm","US2_B_TEXm","US2_C_TEXm",
             "US2_A_TEXp","US2_B_TEXp","US2_C_TEXp",
             "US3_mgm",
             "US3_A_PROseq","US3_B_PROseq","US3_C_PROseq",
             "US3_A_RNAseq","US3_B_RNAseq",
             "US3_A_TEXm","US3_B_TEXm","US3_C_TEXm",
             "US3_A_TEXp","US3_B_TEXp","US3_C_TEXp"
             )

p1 <- plot_bracken(file = bdat,
                   top = 4,
                   pal = "a30000,f5b841,00798c,76e5fc,9acd32,8b008b,66cdaa",
                   otherColor = "gray",
                   sample_order = samples)

p1 <- p1 + geom_vline(xintercept = 12.5, color = "black")
p1
```

## ggsave
```{r}
setwd("/workdir/users/acv46/stool_PROSeq3/figures/kraken/")
ggsave(plot = p1,
       filename = "kraken_phylum_relative_abudance_aligned_v1.pdf",
       width = 7,
       height = 4,
       units = "in",
       device = cairo_pdf)
```

# create plot for aligned reads -- new vs old data
```{r}
setwd("/workdir/users/acv46/stool_PROSeq3/kraken/merged_bracken_bams_ALL")

notlogical <- cols(taxonomy_lvl=col_character())
bdat <- read_tsv("bracken_merged_F.txt", col_names = TRUE, col_types = notlogical) 
  mutate(name = ifelse(name == "Firmicutes", "Bacillota", name),
         name = ifelse(name == "Bacteroidetes", "Bacteroidota", name),
         name = ifelse(name == "Actinobacteria", "Actinomycetota", name)) |>
  select(-contains("TEX"))
samples <- c("old_US2_mgm",
             "old_US2_A_PROseq","old_US2_B_PROseq",
             "old_US2_A_RNAseq","old_US2_B_RNAseq",
             "US2_mgm",
             "US2_A_PROseq","US2_B_PROseq","US2_C_PROseq",
             "US2_A_RNAseq","US2_B_RNAseq",
             "old_US3_mgm",
             "old_US3_A_PROseq","old_US3_B_PROseq",
             "old_US3_A_RNAseq","old_US3_B_RNAseq",
             "US3_mgm",
             "US3_A_PROseq","US3_B_PROseq","US3_C_PROseq",
             "US3_A_RNAseq","US3_B_RNAseq"
             #"US2_A_TEXm","US2_B_TEXm","US2_C_TEXm",
             #"US2_A_TEXp","US2_B_TEXp","US2_C_TEXp",
             #"US3_A_TEXm","US3_B_TEXm","US3_C_TEXm",
             #"US3_A_TEXp","US3_B_TEXp","US3_C_TEXp"
             )

p1 <- plot_bracken(file = bdat,
                   top = 11,
                   #pal = "a30000,00798c,76e5fc,f5b841,9acd32,8b008b,66cdaa",
                   pal = "5c2751,ef798a,f7a9a8,00798c,6457a6,9dacff,76e5fc,a30000,ff7700,f5b841,9acd32",
                   otherColor = "gray",
                   sample_order = samples)

p1 <- p1 + geom_vline(xintercept = 11.5, color = "black") +
           geom_vline(xintercept = 5.5, color = "black", linetype = "dashed") +
           geom_vline(xintercept = 16.5, color = "black", linetype = "dashed")
p1
```

## ggsave
```{r}
setwd("/workdir/users/acv46/stool_PROSeq3/figures/kraken/")
ggsave(plot = p1,
       filename = "kraken_family_relative_abudance_aligned_oldVnew_v1.pdf",
       width = 8,
       height = 5,
       units = "in",
       device = cairo_pdf)
```